### ISSUE TITLE
Consider Step Forward hotkey?

### ISSUE THREAD

**Issue by terracoda** [phetsims/molecules-and-light#336]
The Pause/Play hotkey worked great, and one comment was it would be good to have one for Step Forward, too.

I am wondering if we should consider this for future releases, and for the Timing Control widget, in general.

**Comment by terracoda** [comment]
Decision: Let's try ALT +  L for Step Forward.
@terracoda to update Keyboard dialog content.

**Comment by jessegreenberg** [comment]
@terracoda said
>JUST had a thought that for non-visual exploration bigger steps would be way more useful. @jessegreenberg would it be possible to make ALT + L do big steps instead of little ones? I'm thinking 3 visual steps for photon to reach molecule:
 >- Step one: leaves light source
 >- Step two: halfway to molecule
 >- Step three: at absorption point of molecule

**Comment by jessegreenberg** [comment]
The problem with such large steps is that there is a small distance around the molecule to query for absorption. The center of the molecule needs to enter this rectangle to test for absorption
![image](https://user-images.githubusercontent.com/6396244/81105177-89dd7800-8ee1-11ea-9a33-98d11b5df87c.png)

If the time step is too large, it would be possible for the photon to step completely over the molecule and never be absorbed.

**Comment by jessegreenberg** [comment]
The above commit adds the hot key to molecules and light with the same behavior as the step forward button.

**Comment by jessegreenberg** [comment]
I added a query parameter `keyboardTimeStep` (in seconds) that lets you change the time step for "Alt + L" key press. The default value is 0.0167.

https://phet-dev.colorado.edu/html/molecules-and-light/1.5.0-dev.49/phet/molecules-and-light_en_phet.html?keyboardTimeStep=0.0167.

@terracoda @Matthew-Moore240 can you try it out to see if we should explore this further?

**Comment by jessegreenberg** [comment]
Over slack @terracoda said
> Are steps measured in pixels only? Is there no way to make that fourth step always hits the absorption point?

This would be much more complicated, and currently isn't supported by the model. The model steps forward in time but this would involve moving the photon to a new position independent of time.

**Comment by terracoda** [comment]
@jessegreenberg 0.0835 seems to work really well. It gives about 5 steps instead of 25 or 26.

**Comment by terracoda** [comment]
I'd need to review the description and sound a little more carefully.

**Comment by terracoda** [comment]
@jbphet and @Ashton-Morris could you have a listen to the MAL with parameter set to 0.0835

https://phet-dev.colorado.edu/html/molecules-and-light/1.5.0-dev.49/phet/molecules-and-light_en_phet.html?keyboardTimeStep=0.0835

Then do the following:
1. Turn photons on
2. Use ALT + K to Pause action
3. Use ALT + L to step forward action in big (bigger) steps

I think it sounds really good, but it would be nice for you sound experts to have a combined look/listen.

**Comment by terracoda** [comment]
In my opinion the descriptions and sounds work well with the bigger step size.

And for the record, the proposition in this issue is to make the ALT + L hotkey use the bigger step, and not change the default step size that happens with the Step Forward button.

**Comment by terracoda** [comment]
The two hotkeys (ALT + K and ALT + L) work really nicely together. I can leave the my focus on one of the radio groups and then pause and play at will and choose to step forward at will.

I'm finding the bigger steps really useful, so once sound timing is verified I think we should add the ALT + L hotkey.

**Comment by jessegreenberg** [comment]
I agree, that step size feels nice! A couple of times I was able to get the photon to completely step over the molecule's absorption area (and so we didn't get a "photon passes through" alert), but it is rare.

**Comment by jbphet** [comment]
In general, I think the behavior of the sound in this case seems reasonable.  I do have some questions based on using the sim this way (something that I don't do very often but perhaps should):

1. When switching from play to pause and vice versa using the alt-k hot key, the play/pause sound is not played.  If I use tab navigation and switch them using focus+space, the sound *is* played.  Is it intentional not to play the sound in this case?
1. The same goes for the alt-L interaction - it doesn't produce the step sound.
1. Holding down alt-L doesn't produce continuous sequential steps, and I found myself wanting this.  Is this something we should consider adding?
1. This isn't directly related to the hot keys, but I noticed while testing that if I press and hold the step button with the mouse it produces continuous sequential steps, but it only produces the press sound for the initial press, which seems right to me.  If I do the same thing using tab navigation and the Enter key, the press sound is produced for every step, which is pretty obnoxious.  Should we consider changing this behavior?

**Comment by terracoda** [comment]
Great questions @jbphet 
1 & 2 - I assume the sound should play the same when pausing/playing and stepping. I didn't specifically request this, so glad you asked. I think we might need sound listeners for the custom keys as they are custom. Do you need to do that or can @jessegreenberg?
3. Custom shortcuts like this are new territory. I wouldn't want to do anything like press and hold until we know more. The ALT + L is meant to be more akin to the Space key behaviour than Enter key behaviour. 
4. Hmm, I am assuming this happens because there are different kinds of events on the button with these different interactions. The Enter key is a press with auto-release. The Space key is just a press, and you have to physically release it to get a release event. A mouse press-and-hold is likely different, again. @jessegreenberg can you comment on this? You have explained it to me before. Regarding the sound, I haven't had any complaints about the sound repeating with press-and-hold from BVI participants. This is our first Step Forward button, I assumed it was meant to be like that from the beginning since there are multiple steps happening. I never compared the mouse press-and-hold directly with the Enter key press-and-hold. I was too focused on good context responses for that interaction. @jessegreenberg, even if we wanted to only have the sound play once, is that even possible?

**Comment by jessegreenberg** [comment]
For the sound, this isn't something I have done before but I would be happy learn! @jbphet let me know what you think is best here.

> Holding down alt-L doesn't produce continuous sequential steps, and I found myself wanting this.

As a data point, web video/audio players scrub when you press and hold arrow keys. Our usage is a bit different though.

> If I do the same thing using tab navigation and the Enter key, the press sound is produced for every step, which is pretty obnoxious.

That is true, and will happen for all of PhET's buttons (ResetAllButton, PlayPauseButton, sound button, others). The reason this happens is because we get fewer DOM events when a screen reader is in use. Normally we get `keydown` and `keyup` events like we get `down` and `up` events with mouse/finger. But with a screen reader we **only** get a single `click` event. So we cannot play a single sound on `keydown` because we don't reliably get that event. And we have now way to reliably distinguish between a user pressing "enter" multiple times vs pressing and holding the button down. But maybe it could be improved by only playing a sound once every time interval from keyboard input (or something).

**Comment by jbphet** [comment]
> For the sound, this isn't something I have done before but I would be happy learn! @jbphet let me know what you think is best here.

No time like the present!  So yes @jessegreenberg , it would probably be good for you try adding the sound, since we need to get all developers a bit more familiar with it.  There are a set of singleton sound instances that can be included in the appropriate file that should, I think, make this pretty straightforward.  Please have a look at `playSoundPlayer`, `pauseSoundPlayer`, and `stepForwardSoundPlayer`, as well as their usages in the code base, and go from there.  I'm always available to consult if needed.

**Comment by terracoda** [comment]
@jessegreenberg, if 0.0666 is more reliable as you say in https://github.com/phetsims/molecules-and-light/issues/336#issuecomment-624751344, I am good with that timing.

0.0666 still results in about 5 or 6 bigger steps rather than the 25 or 26 little by little steps one gets by default with the Step Forward button.

**Comment by terracoda** [comment]
As for if the sound should fire repeatedly when holding down Enter (or ALT + L). I think some regular repetition of the sound when a Step Forward interaction is happening benefits non-visual users because we do not describe the photons progression across the screen.

**Comment by terracoda** [comment]
@jessegreenberg, both @Matthew-Moore240 and I agree the "bigger" Step Forward step with ALT + L is really nice. Assigning to you for implementation.

I am not sure if it is best to continue the discussion on Sound here or in a new issue?

Please create a new issue if that makes sense.

**Comment by terracoda** [comment]
@jessegreenberg, again, for the sound issue, as you said the same repeat sound happens when holding down any button (Including the very strong sounding Reset All button).

What are our options for optimization here, and I am thinking this kind of optimization is a general one, one that would be what we want for all sims, not just MAL.

**Comment by terracoda** [comment]
@jbphet, @emily-phet, and @Ashton-Morris does this issue warrant a sound meeting discussion?
The is about repeat firings of sound when a button is held down with the Enter Key.

I have a feeling this has come up before?

**Comment by jessegreenberg** [comment]
>  if 0.0666 is more reliable as you say in #336 (comment), I am good with that timing.

OK sounds good, thanks!

>  would probably be good for you try adding the sound, since we need to get all developers a bit more familiar with it.

Cool, sounds great! Ill give it a try, and @terracoda ill open a new issue to track this.

> What are our options for optimization here, and I am thinking this kind of optimization is a general one, one that would be what we want for all sims, not just MAL.

Hmm, I can think of three options at the moment but there may be more.
- Limit button sounds to only play once every X ms.
- Prevent all button sounds if we get click events faster then once every X ms.
- Limit button presses to only be able to fire once every X ms with a keyboard.

**Comment by jessegreenberg** [comment]
The time step for the global hotkey has been set.
> does this issue warrant a sound meeting discussion?

I would be happy to join too, this issue involves both alternative input and sound.

**Comment by emily-phet** [comment]
@terracoda looped me in re repeating button sound. 

I doubt there is a general solution for this - "button" is a big category within PhET and depending on the context it may make sense to hear repeating sounds or to hear only one etc. 

If you all think this is a major issue for MAL, I suggest it be handled as a sim specific one. If it's a minor thing, I suggest leaving it as is to not delay publication.

Unfortunately, I don't have the bandwidth to participate in design solutions, but if you all decide to address this for MAL, please loop @jbphet and @Ashton-Morris in for at least one quick conversation.

**Comment by jbphet** [comment]
@terracoda said:

>  I think some regular repetition of the sound when a Step Forward interaction is happening benefits non-visual users because we do not describe the photons progression across the screen.

...and @emily-phet said:

>  If it's a minor thing, I suggest leaving it as is to not delay publication.

Even though I brought it up, my vote now would be that, given the input above, we leave it as is and revisit if it ever turns out to be a problem during interviews or if we get any complaints from users.

**Comment by Ashton-Morris** [comment]
> @jbphet and @Ashton-Morris could you have a listen to the MAL with this keyboard Time Step parameter set to 0.0835
> 
> https://phet-dev.colorado.edu/html/molecules-and-light/1.5.0-dev.49/phet/molecules-and-light_en_phet.html?keyboardTimeStep=0.0835
> 
> Then do the following:
> 
> 1. Turn photons on
> 2. Use ALT + K to Pause action
> 3. Use ALT + L to step forward action in big (bigger) steps
> 
> I think it sounds really good, but it would be nice for you sound experts to have a combined look/listen.

I think it sounds really good. I would be open to a short sound related meeting to address what you are describing above

**Comment by terracoda** [comment]
@Ashton-Morris, glad you agree.
I think we nailed it. Please see issue #339 which I just closed.

We are going to wait on the "repeating sound" part of the issue.

I think this issue is all verified, so closing.
